### PR TITLE
Add TaggedScheduler, couple lifetime of tasks and orchestrator

### PR DESCRIPTION
**Pull request by mwilsnd**
This PR refactors the scheduler and thread pool architecture by coupling scheduled tasks with an opaque identifier intended to represent the owner of the submitted task. This enables queries on the thread pool based on tasks submitted by a specific owner, which is relevant as the same background thread pool can be shared among multiple map instances. Knowing what tasks are scheduled by a specific owner allows us to remove any timeout conditions when waiting on all submitted work to be completed. This should result in more consistent and deterministic behavior of background tasks in multiple map scenarios.

### CODE REVIEW

**Comment by louwers** [comment]
Should probably undo https://github.com/maplibre/maplibre-native/pull/2455

**Comment by TimSylvester** [comment]
> Should probably undo #2455

It'll disappear in the merge, that code was removed.

**Comment by louwers** [comment]
image

**Review by TimSylvester — APPROVED** [review]

**Review by louwers — COMMENTED** [review]
From my testing it works great.

Having a `void*` in a C++ API feels intuitively wrong (even though it's just used as an identifier) and may be flagged by code quality tools at some point.  Have you considered other approaches?

**Comment by TimSylvester** [comment]
`SimpleIdentity` would work

**Comment by louwers** [comment]
> SimpleIdentity

I like that.

**Review by louwers — APPROVED** [review]

**Review by TimSylvester — APPROVED** [review]

File: src/mbgl/tile/tile_cache.cpp [hunk 1]
```diff
@@ -51,9 +51,15 @@ void TileCache::deferredRelease(std::unique_ptr<Tile>&& tile) {
     // If this temporary outlives the `schedule` call, and the function is executed immediately
     // by a waiting thread and is already complete, that temporary reference ends up being the
     // last one and the destruction actually occurs here on this thread.
-    std::function<void()> func{[tile_{CaptureWrapper<Tile>{std::move(tile)}}]() {
+    std::function<void()> func{[tile_{CaptureWrapper<Tile>{std::move(tile)}}, this]() mutable {
+        std::exchange(tile_.item, {});
```
Comment (louwers) [inline src/mbgl/tile/tile_cache.cpp:55]: Why not `tile_.item = {}`?

File: test/util/thread.test.cpp [hunk 2]
```diff
@@ -434,8 +420,55 @@ TEST(Thread, WrongThread) {
...
+
+    // Do the same for queue 2
+    stopTasks2 = true;
+    poolTag2.waitForEmpty();
+    const auto totalRuns2 = runCount2.load();
+
+    // No other tasks in queue 1 ran while waiting on queue 2.
+    ASSERT_TRUE(totalRuns1 == runCount1);
+
+    std::this_thread::sleep_for(std::chrono::milliseconds(150));
```
Comment (louwers) [inline test/util/thread.test.cpp:470]: Is this needed?

Comment (louwers) [reply test/util/thread.test.cpp:470]: Ah OK you make sure that after `.waitForEmpty()` no other tasks run, got it.

File: platform/android/MapLibreAndroid/src/cpp/map_renderer.hpp [hunk 3]
```diff
@@ -119,7 +123,8 @@ class MapRenderer : public Scheduler {
     float pixelRatio;
     std::optional<std::string> localIdeographFontFamily;
 
-    std::shared_ptr<ThreadPool> threadPool;
+    // std::shared_ptr<ThreadPool> threadPool;
```
Comment (louwers) [inline platform/android/MapLibreAndroid/src/cpp/map_renderer.hpp:126]: stray comment

Comment (louwers) [inline platform/android/MapLibreAndroid/src/cpp/map_renderer.hpp:126]:
```suggestion
```

**Review by louwers — APPROVED** [review]
Very clean.

**Review by TimSylvester — APPROVED** [review]

File: src/mbgl/tile/tile_cache.cpp (hunk 1 above)
Comment (mwilsnd) [reply src/mbgl/tile/tile_cache.cpp:55]: I was debugging before I remembered I needed to make this lambda mutable.

**Comment by westnordost** [comment]
If there will be a new release or beta/alpha release with this change, we could test if it solves #2410

**Comment by louwers** [comment]
@westnordost I made a pre-release. https://github.com/maplibre/maplibre-native/releases/tag/android-v11.0.2-pre0

**Comment by westnordost** [comment]
Thank you, I already added it to the branch and asked if it can still be reproduced: https://github.com/streetcomplete/StreetComplete/pull/5693#issuecomment-2206570705

**Pull request merged**